File: JavaToKotlin/app/src/main/java/io/dustin/apps/board/api/usecase/qna/question/DeleteQuestionUseCase.java
package io.dustin.apps.board.api.usecase.qna.question;

import io.dustin.apps.board.domain.qna.question.model.Question;
import io.dustin.apps.board.domain.qna.question.model.dto.QuestionDto;
import io.dustin.apps.board.domain.qna.question.service.ReadQuestionService;
import io.dustin.apps.board.domain.qna.question.service.WriteQuestionService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.security.Principal;

@Service
@RequiredArgsConstructor
public class DeleteQuestionUseCase {

    private final ReadQuestionService readQuestionService;
    private final WriteQuestionService writeQuestionService;

    @Transactional
    public QuestionDto execute(Long userId, Long questionId) {
        Question question = readQuestionService.findById(questionId);
        if(!question.getUserId().equals(userId)) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "삭제 권한이 없습니다.");
        }
        writeQuestionService.delete(question);
        return QuestionDto.from(question);


    }


}

File: apps/src/main/kotlin/io/dustin/apps/board/api/usecase/qna/question/DeleteQuestionUseCase.kt
package io.dustin.apps.board.api.usecase.qna.question

import io.dustin.apps.board.api.qna.request.command.QuestionDeleteCommand
import io.dustin.apps.board.domain.qna.question.model.dto.QuestionDto
import io.dustin.apps.board.domain.qna.question.service.ReadQuestionService
import io.dustin.apps.board.domain.qna.question.service.WriteQuestionService
import io.dustin.apps.common.exception.DataNotFoundException
import io.dustin.apps.common.model.response.CommonResponse
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import javax.naming.AuthenticationException

@Service
class DeleteQuestionUseCase (
    private val readQuestionService: ReadQuestionService,
    private val writeQuestionService: WriteQuestionService
) {
    @Transactional
    fun execute(command: QuestionDeleteCommand) {
        val question = readQuestionService.findByIdOrThrow(command.questionId)
        if(question.userId == command.userId){
            writeQuestionService.delete(question)
        }
        else{
            return throw AuthenticationException("권한이 없습니다.")
        }

    }
}
